[dart] fix some compilation issues and added casts for analyzer implicit-casts flag

Grohden opened the pull request:
### Details

I've made some fixes for the **dart** templates, @kuhnroyal and I had some discussion on #8179.

My changes were necessary to be able to compile **my personal project** and also the fake petstore generated clients.

**Not all changes were made to fix compile errors**: I've mistakenly enabled the `implicit-casts:false` on the generated client and then fixed the resulting analyzer errors.. 
 * if fixing casts 'issues' and supporting that flag is not desired I can revert my cast changes.
 * if fixing casts 'issues' and supporting that flag is desired we could enable the strong mode like @kuhnroyal did in #8231

From the changelist listed in #8179 I've fixed:
  - [x] De-serialization of maps of maps e.g. Map<String, Map<String, String>>

(I've only fixed the compilation error, not sure if it will work properly on runtime)

Maybe I've fixed other issues.. but I'm not sure, would be glad if someone test it and confirm exactly what is still broken

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) beforehand.
- [x] Run the shell script `./bin/generate-samples.sh`to update all Petstore samples related to your fix. This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. These must match the expectations made by your contribution. You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`
- [x] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

@kuhnroyal could you comment on the casts you said that were uneeded?

@swipesight @jaumard @josh-burton @amondnet @sbu-WBT @kuhnroyal @agilob

agilob commented on modules/openapi-generator/src/main/resources/dart2/class.mustache:
```diff
@@ -190,22 +188,22 @@ class {{{classname}}} {
-    if (json != null && json.isNotEmpty) {
-      json.forEach((String key, dynamic v) {
-        map[key] = {{{classname}}}.listFromJson(v, emptyIsNull: emptyIsNull, growable: growable);
+    if (json?.isNotEmpty == true) {
```
do we need `== true`?

Grohden replied:
yes

```dart
void main() {
  if(null) {
    print("x");
  }
}

// Uncaught Error: Assertion failed: "boolean expression must not be null"
```

kuhnroyal replied:
I understand this change but in any case we don't care about wether it is empty. `foreach` doesn't run if it is empty.
Maybe something like this but thats just my preference, up  to you.
```dart
return json?.map((key, dynamic value) => MapEntry(key, {{{classname}}}.listFromJson(value, emptyIsNull: emptyIsNull, growable: growable,),) ?? <String, {{{classname}}}>{};
```

kuhnroyal replied:
Ideally you could change this in another PR as it is probably related to some analyzer rule.

agilob commented on modules/openapi-generator/src/main/resources/dart2/class.mustache:
```diff
@@ -190,22 +188,22 @@ class {{{classname}}} {
-        map[key] = {{{classname}}}.listFromJson(v, emptyIsNull: emptyIsNull, growable: growable);
+    if (json?.isNotEmpty == true) {
+      json.forEach((key, value) {
+        map[key] = {{{classname}}}.listFromJson(value as List<dynamic>, emptyIsNull: emptyIsNull, growable: growable,);
```
can you assess if this has any performance hit? Can casting things inside a loop slower be slower than casting upfront in loop definition?

Grohden replied:
> can you assess if this has any performance hit?

I lack the knowledge for that...

> Can casting things inside a loop slower be slower than casting upfront in loop definition?

U mean to move this to the method signature? declaring `Map<String, dynamic> json` as `Map<String, List<dynamic>> json`?

On that note.. isn't this a microoptimization? I know that dart runtime will validate all assignment types whether you cast or not something, maybe the compiler make some optimizations when you cast before to not validate it again :thinking:... but again, I lack the knowledge for that.

oh, and [json_serializable](https://pub.dev/packages/json_serializable) also generates theses casts that I've added

kuhnroyal replied:
Casts usually don't have a runtime impact, they are mostly there to make the compiler happy.

agilob commented on modules/openapi-generator/src/main/resources/dart2/api.mustache:
```diff
@@ -192,7 +192,7 @@ class {{{classname}}} {
       {{^isArray}}
         {{#isMap}}
-      return {{{returnType}}}.from(apiClient.deserialize(_decodeBodyBytes(response), '{{{returnType}}}'));
+      return {{{returnType}}}.from(apiClient.deserialize(_decodeBodyBytes(response), '{{{returnType}}}') as Map<dynamic, dynamic>);
```
we already know it's a map of things, what does knowing it is `<dynamic, dynamic>` bring?

Grohden replied:
With the implicit casts false we get this:
![image](https://user-images.githubusercontent.com/7496871/102718391-a4c0f900-42c6-11eb-819d-c43d787fc512.png)

which means that we can't implicit cast a type to a more specialized type (`dynamic` to `Map<dynamic, dynamic>`)

Grohden replied:
maybe we could just cast it instead of applying the map.from? :thinking: gonna try that in the https://dartpad.dev/

Grohden replied:
just casting it would work:
```dart
void main() {
  print(foo({"a": 1}));
}

Map<String, int> foo(dynamic response) {
  return response as Map<String, int>;
}
```

agilob replied:
that's fine I'm cool with it, just was asking

agilob commented on modules/openapi-generator/src/main/resources/dart2/class.mustache:
```diff
@@ -109,15 +109,15 @@ class {{{classname}}} {
           : (json[r'{{{baseName}}}'] as List).map(
       {{#items.complexType}}
-              {{items.complexType}}.listFromJson(json[r'{{{baseName}}}'])
+              {{items.complexType}}.listFromJson(json[r'{{{baseName}}}'] as List<dynamic>)
```
same here, we know that the method accepts `List<dynamic> json` as first param, so why the cast here?

Grohden replied:
same as above, with the `implicit-casts:false` we cant implicit cast `dynamic` to `List<dynamic>`, but as I said in the description, if that flag is not desirable I can remove almost all casts :thinking: and I'm pretty sure it will still work

agilob commented on modules/openapi-generator/src/main/resources/dart2/api_client.mustache:
```diff
@@ -70,7 +70,7 @@ class ApiClient {
     String path,
     String method,
-    Iterable<QueryParam> queryParams,
+    List<QueryParam> queryParams,
```
User of this API should be able to pass `Set` as well, eg. let dart remove duplicated elements and not relay on order of elements in Iterable.

Grohden replied:
I've changed this one based on the call to `_updateParamsForAuth`, this function expect the param to be a list, passing a set would be accepted but would result in runtime error when you call that function

![image](https://user-images.githubusercontent.com/7496871/102719228-a345ff80-42cb-11eb-8afe-f69fe85689c6.png)

```
void main() {
  foo({"a", "b"});
}

void foo(Iterable<String> authNames) {
  bar(authNames);
}

void bar(List<String> authNames) {
}

// Uncaught Error: TypeError: Instance of '_LinkedHashSet<String>': type '_LinkedHashSet<String>' is not a subtype of type 'List<String>'
```

agilob replied:
There doesn't seem to be any reason for `_updateParamsForAuth` being limited to `List` it's not order dependent either

```dart
  /// Update query and header parameters based on authentication settings.
  /// @param authNames The authentications to apply
  void _updateParamsForAuth(List<String> authNames, List<QueryParam> queryParams, Map<String, String> headerParams) {
    authNames.forEach((authName) {
      Authentication auth = _authentications[authName];
      if (auth == null) throw new ArgumentError("Authentication undefined: " + authName);
      auth.applyToParams(queryParams, headerParams);
    });
  }
```

Set seems to be even more correct options as some frameworks (I remember one PHP based doing it) will throw error if you pass the same url param twice.

Grohden replied:
Well, what you said make sense, I'm gonna try to change the signature of `_updateParamsForAuth` and `applyToParams` to receive a `Iterable` instead of a `List` then

Grohden replied:
I forgot to thank you for the review, thanks for that :heart: 

It turns out that I can't change the signature of `applyToParams` to a `Iterable`: in `ApiKeyAuth` there's a call to `.add`, and Iterables don't have that method :/

We could change the method and add a `.toList()` call:
```dart
  void _updateParamsForAuth(
    List<String> authNames,
    Iterable<QueryParam> queryParams,
    Map<String, String> headerParams,
  ) {
    for(final authName in authNames) {
      final auth = _authentications[authName];
      if (auth == null) {
        throw ArgumentError('Authentication undefined: $authName');
      }
      auth.applyToParams(queryParams.toList(), headerParams);
    }
  }
```

But I'm pretty sure that this would make the `applyToParams` implementers `.add` calls pointless.. they rely on mutability, and `.toList` would create a copy of the original list

agilob replied:
Does that mean it currently throws at runtime if we pass Set in place of the existing iterable, which at the end is expected to be a List?

agilob replied:
so basically for now it has to be specific implementation Set or List but not parent class `Iterable` because it doesn't have `.add` method :clap:. 

sure, leave it as list then

agilob replied:
>We could change the method and add a .toList() call:

My personal choice would be to pass iterable to allows sets and any other lists and sets, but it's also outside of scope of your PR, so I'm leaving the decision with you.

Grohden replied:
I'm gonna keep it as a List, ideally we would use iterable and not rely on mutability (as it may affect code outside of this generated lib).

I'm gonna try to open another PR to make it accept a iterable and also rely on pure functions instead of mutating the list

kuhnroyal replied:
Is this external API? If so, this is a breaking change, 5.0.0 was just released. But it also fixes a bug so I am not sure what the process is here. @agilob You gonna fix this before the next release? :)

kuhnroyal commented:
You should add a matching `analysis_options.yaml` as a new support template file, like it is done in the dio-generator.

agilob commented:
@kuhnroyal  @Grohden 
 what do you two think about adding something more popular and opinionated that a custom analysis? Would you add effective-dart here?

Grohden commented:
@agilob would totally use use effective dart, I can make the needed changes if @kuhnroyal also agrees

kuhnroyal commented:
I would love to see a linter in dart/dart-io at some point.
Need to decide which makes the most sense here.
https://pub.dev/packages/effective_dart
https://pub.dev/packages/lint
https://pub.dev/packages/pedantic

I usually use lint.
And yes, should be done in another PR and added to both generators as well as to the integration tests.

agilob commented:
Pedantic always makes my eyes bleed, it makes Dart have more strict syntax than Java. I personally stick with effective_dart, but I've never used `lint`... and I think I know why ` An opinionated, community-driven set of lint rules for Dart and Flutter projects. Like pedantic but stricter `

agilob commented:
>edit: effective could be done in another PR to not bloat this one

I disagree here, since you're adding lint fixes it makes sense to add the linter too!

Grohden commented:
> I disagree here, since you're adding lint fixes it makes sense to add the linter too!

Somehow this PR ended up as lint fixes, but my primary problem was the `Map.mapfromJson` not letting me compile my project.. I've mistakenly enabled the analyzer and fixed these cast issues :laughing: I can remove the `implicit-casts:false` and its fixes and enable it in a separated PR for the linter if that's better

kuhnroyal commented:
Yes, we should have a separate PR for the strong-mode/implicit changes. Can you use the same options I used here https://github.com/OpenAPITools/openapi-generator/pull/8231
I just checked out this branch and tried with the `analysis_options` from my branch and all the `json as Foo` casts are somehow unnecessary. So I would really love to see your analyzer options.

agilob commented:
No probs, so just add the analysis.yaml so we know what the changes are related to.

Grohden commented:
@kuhnroyal  my analysis options for the petstore is just:
```yaml
analyzer:
  strong-mode:
    implicit-casts: false
```

kuhnroyal commented:
Ok so I see the problem now. I never disable implicit casts because that doesn't do much good.
Assigning `dynamic` to a `String` is just as safe as casting it, if it is not a `String` it will fail either way.

On the other hand having `implicit-dynamic: false` actually prevents type errors.
After realizing we are talking about different settings I have to say I am against `implicit-casts: false`.

kuhnroyal commented on modules/openapi-generator/src/main/resources/dart2/class.mustache:
```diff
@@ -133,17 +133,15 @@ class {{{classname}}} {
-        {{{name}}}: json[r'{{{baseName}}}'] == null
-          ? null
-          : {{{complexType}}}.mapFromJson(json[r'{{{baseName}}}']),
+        {{{name}}}: json[r'{{{baseName}}}'] as {{{datatype}}},
```
Are we sure that this can only be a primitive type here?

Grohden replied:
This is the map of maps case...

I don't see this case in the petstore, but my project had this case:
![image](https://user-images.githubusercontent.com/7496871/102776642-db9e1a00-436d-11eb-88a7-5e7727b8c93c.png)

I'm gonna change this line to the same as the list treatment (above)
```mustache
          {{#items.isMap}}
            {{{name}}}: json[r'{{{baseName}}}'] == null
              ? null
                {{#items.complexType}}
                    : {{items.complexType}}.mapFromJson(json[r'{{{baseName}}}']),
                {{/items.complexType}}
                {{^items.complexType}}
                    : (json[r'{{{baseName}}}'] as Map).cast<String, Map>(),
                {{/items.complexType}}
          {{/items.isMap}}
          {{^items.isMap}}
              {{{name}}}: json[r'{{{baseName}}}'] // primitive? Probably Map<String, primitive>
          {{/items.isMap}}
```

This is what it will generate:
![image](https://user-images.githubusercontent.com/7496871/102778165-d55d6d00-4370-11eb-9447-a5da40c3f8ca.png)

Its that ok?

Grohden commented:
@kuhnroyal @agilob 
I've removed the fixes related to the `implicit-casts: false`, I've also changed the way I deal with the `Maps of maps` case

> After realizing we are talking about different settings I have to say I am against implicit-casts: false.

I've only learned from your comments about the `implicit-dynamic: false`, so agree with you and my recent changes are respecting this flag, but I'm gonna open another PR for the analysis file with it.

agilob reviewed (changes requested):
Can you add the analysis yaml file pls?

Grohden commented:
@agilob 
U mean the one from #8231 or
one with
```
analyzer:
  strong-mode:
    implicit-casts: true
```
(which is to **explicitly accept** `implicit-casts`)
?

agilob commented:
The one that triggers warnings you fixed here

Grohden commented:
@agilob file added

Grohden commented:
@kuhnroyal I was bugged about why I had analyzer errors reported on the `_updateParamsForAuth` from the `invokeAPI`, it turns out that it was the `implicit-casts: false` that was reporting that error to me.

So just so you know, this code exemplifies a case where this flag can be helpful:
```dart
void main() {
  foo({"a", "b"});
}

void foo(Iterable<String> authNames) {
  // implicit-casts: false -> The argument type 'Iterable<String>' can't be assigned to the parameter type 'List<String>'.
  // implicit-casts: true -> ok, just runtime error :D
  bar(authNames); 
}

void bar(List<String> authNames) {
  authNames.add("bar");
}
```

agilob commented:
>>  // implicit-casts: false -> The argument type 'Iterable<String>' can't be assigned to the parameter type 'List<String>'.

Yea, that's a good fix. I'll address the bigger issue some time later, it shouldn't be List there :(

kuhnroyal commented:
I agree that this is one of the few cases were `implicit-cast: false` would highlight the potential bug. The question is, is the overhead in all the other places worth it? I am not sure. What do you guys think?

kuhnroyal commented on modules/openapi-generator/src/main/resources/dart2/api_client.mustache:
```diff
@@ -209,12 +209,12 @@ class ApiClient {
     Map<String, String> headerParams,
   ) {
-    authNames.forEach((authName) {
+    for(final authName in authNames) {
```
Missing space

agilob replied:
where is the space missing? It's all good.

kuhnroyal replied:
`for (`. Whatever, we need to add post-processing to the generated examples so we have diffs that don't have this problem.

kuhnroyal commented on modules/openapi-generator/src/main/resources/dart2/enum.mustache:
```diff
@@ -78,4 +78,4 @@ class {{{classname}}}TypeTransformer {
   static {{{classname}}}TypeTransformer _instance;
-}
\ No newline at end of file
+}
```
Remove the newline

agilob replied:
It's all good, this is autogenerated code, shouldn't even be committed to version control, there are plenty of generators that leave +5 empty whitespaces at end of files.

Grohden replied:
These new lines were added by AS, but I'm with @agilob on this one.. I always apply dartfmt on the generated client so things like empty lines are removed.. but when we add the effective dart guide the generated code will need to be properly formatted when generated isn't?

kuhnroyal replied:
The problem is that it will change again for the next PR and the next and the next. That is lots of noise that doesn't actually affect the commit. And both enum templates are actually inlined so you suddenly end up with duplicate/triple empty lines between classes. But whatever, up to you.

Grohden replied:
would be nice to properly format the **mustaches** templates and leave the dart fonts reformat for the user.. I find very confusing to read if else conditions on the current mustache templates

kuhnroyal commented on modules/openapi-generator/src/main/resources/dart2/enum_inline.mustache:
```diff
@@ -78,4 +78,4 @@ class {{{classname}}}{{{enumName}}}TypeTransformer {
   static {{{classname}}}{{{enumName}}}TypeTransformer _instance;
-}
\ No newline at end of file
+}
```
Remove the newline

Grohden commented:
@kuhnroyal I am the "if the compiler say its ok, then its ok" guy, so after seeing the flag capturing an actual bug I'm not against it - I'm gonna keep using it in personal projects. 

But I agree that having to cast every single dynamic is annoying and unnecessary, and the dart API that we have is not huge and is not expected to change very often.. so I would keep the flag (explicitly) disabled

agilob commented:
>would highlight the potential bug. The question is, is the overhead in all the other places worth it? I am not sure. What do you guys think?

Let's keep what we have here and address linter in another PR. I think it's fair to add the strictest linter in the generated code. I can do it after this gets in.

agilob commented:
>But I agree that having to cast every single dynamic is annoying and unnecessary, and the dart API that we have is not huge and is not expected to change very often.. so I would keep the flag (explicitly) disabled

In case of autogenerated code I think correctness and type-safety is more important than "I dont like it has 2 casts every 5 lines in files I never modify manually"

kuhnroyal commented:
I agree, since it is generated we can add the casts. But they need to be the correct ones and that is not always easy to see.
For the bug that was found, most people would have just blindly added the cast `as List` to make it compile. It would still fail with a `Set` but the compiler is happy.

wing328 commented:
@Grohden can you please resolve the merge conflicts when you've time? Thank you.

Grohden commented:
@wing328 I've merged the latest master today and regenerated the needed files..

I probably need to test if the new added anaysis_options.yaml file will affect the new (?) json serializable option... but that's gonna take time if it depends on me

agilob commented:
>I probably need to test if the new added anaysis_options.yaml file will affect the new (?) json serializable option...

I can look into it after this one is meged

wing328 commented:
@Grohden sorry do you mind resolving the merge conflicts again? We'll put other Dart PRs on hold for the time being.

Grohden commented:
@wing328 no problem

I've merged today's master and re-generated all dart codes (`./bin/generate-samples.sh ./bin/configs/dart*`)

wing328 commented:
Travis CI failure not related to this PR.
